preprocessing.py

```diff
@@ -1,166 +1,180 @@
+    for i in range(len(df)):
+        if re.match(bad_pattern, df['text'][i], flags=re.IGNORECASE):
+            bad_pages.append(i)
+    df=df.drop(labels=bad_pages,axis=0)
+    df=df.reset_index(drop=True)
+    return df
+
+
+#conversion of brands to numericals
+def one_hot_encoding(df):
+    one_hot = pd.get_dummies(df.brand_name)
+    df = pd.concat([df, one_hot], axis=1)
+    return df
+
+
+#get the domain name
+def get_top_level_domain(url):
+    tld = urlparse(url)
+    return tld.netloc
+
+
+#get the domain name's length
+def get_url_len(url_str):
+    return len(url_str)
+
+
+#get n-grams for each text(in our case it is n=2)
+def get_english_char_bigram_probs():
+    words = nltk.corpus.words.words('en') # loading in a corpus of words
+    bg_cnts = {}
+    ung_cnts = {}
+    # going through each word in the corpus
+    for word in words:
+        padding = ' ' # needed for bigram models
+        word = padding + word.lower() + padding
+        # getting characters
+        chars = [c for c in word]
+        bgs = nltk.bigrams(chars) # getting bigrams
+        
+        # adding bigram to counts dictionary
+        for bigram in bgs:
+            bg_cnts[bigram] = bg_cnts[bigram] + 1 if bigram in bg_cnts else 1
+            
+        #adding unigram to count dictionary
+        for unigram in chars:
+            ung_cnts[unigram] = ung_cnts[unigram] + 1 if unigram in ung_cnts else 1
+    
+    # turning each count into a probability (MLE)
+    for k, v in bg_cnts.items():
+        first_word, _ = k
+        denom = ung_cnts[first_word]
+        bg_cnts[k] = v/denom # now it is a conditional probability
+        
+    # same for unigrams
+    norm = sum(ung_cnts.values())
+    for k, v in ung_cnts.items():
+        ung_cnts[k] = v/norm
+
+    return bg_cnts, ung_cnts
+
+
+#score of each url texts' bigrams
+def get_score(ung_probs, bg_probs, url_str):
+    url_chars = [c for c in url_str]
+    url_bgs = nltk.bigrams(url_chars)
+    
+    # getting a score of each url char bigram
+    score = 0
+    for bigram in url_bgs:
+        # using linear interpolatation smoothing in negative log space to prevent underflow (lambda 0.5)
+        first_word, _ = bigram
+        score += 0.5*bg_probs.get(bigram, 0) + 0.5*ung_probs.get(first_word, 0)
+        # TODO probably should do entropy here
+    return score
+
+
+#normal splitting of the dataset into training and testing
+def split_data(df):
+    y = df['status']
+    X = df.drop(['status'], axis=1)
+    X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.1, random_state=42)
+    return X_tr, X_te, y_tr, y_te
+
+
+#store csv's
+def store_datasets(Xtrain,Xtest,ytrain,ytest):
+    Xtrain.to_csv('x_train.csv',index=False)
+    Xtest.to_csv('x_test.csv',index=False)
+    ytrain.to_csv('y_train.csv',index=False)
+    ytest.to_csv('y_test.csv',index=False)
```

Comment:
change the dir to be 'data/FILE_NAME' for all of these. Better to store them under data/, I think